File: frontend/src/lib/flows.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { apiDelete, apiGet, apiPatch, apiPost, apiPut } from "./apiClient";
import type { FlowGraph } from "./flowGraph";

/** A saved flow summary (F08): the list shape, without the graph. */
export interface FlowSummary {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
}

/** A saved flow with its full graph — the shape returned on open/save. */
export interface Flow extends FlowSummary {
  graph: FlowGraph;
}

/** Save (create) / re-save (update) request body. */
export interface FlowInput {
  name: string;
  graph: FlowGraph;
}

interface FlowsResponse {
  flows: FlowSummary[];
}

/** Query key for the caller's saved-flows list. */
export const flowsQueryKey = ["flows"] as const;

/** Query key for a single opened flow. */
export function flowQueryKey(id: string) {
  return ["flows", id] as const;
}

/** List the caller's saved flows (`GET /flows`), most recently updated first. */
export function useFlows() {
  return useQuery({
    queryKey: flowsQueryKey,
    queryFn: () => apiGet<FlowsResponse>("/flows").then((data) => data.flows),
  });
}

/**
 * Open a single flow with its full graph (`GET /flows/{id}`). Disabled until an
 * id is provided, so the page can hold `null` when nothing is open.
 */
export function useFlow(id: string | null) {
  return useQuery({
    queryKey: flowQueryKey(id ?? ""),
    queryFn: () => apiGet<Flow>(`/flows/${id}`),
    enabled: id != null,
  });
}

/** Save a new flow (`POST /flows`); invalidates the flows queries on success. */
export function useCreateFlow() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: FlowInput) => apiPost<Flow>("/flows", input),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: flowsQueryKey }),
  });
}

/** Re-save the open flow (`PUT /flows/{id}`); invalidates the flows queries. */
export function useUpdateFlow() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, input }: { id: string; input: FlowInput }) =>
      apiPut<Flow>(`/flows/${id}`, input),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: flowsQueryKey }),
  });
}

/** Rename a flow (`PATCH /flows/{id}`); invalidates the flows queries. */
export function useRenameFlow() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, name }: { id: string; name: string }) =>
      apiPatch<FlowSummary>(`/flows/${id}`, { name }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: flowsQueryKey }),
  });
}

/** Delete a flow (`DELETE /flows/{id}`); invalidates the flows queries. */
export function useDeleteFlow() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => apiDelete<{ id: string }>(`/flows/${id}`),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: flowsQueryKey }),
  });
}

File: backend/src/routes/flows.rs
//! Flows CRUD + rename endpoints (F08), protected by `require_auth` and scoped
//! to the authenticated caller. Each handler pulls `AuthUser`, delegates to the
//! service, and renders the platform success envelope. List returns lightweight
//! summaries (no graph); open returns the full graph.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::Json;
use serde_json::{json, Value};
use uuid::Uuid;

use crate::auth::AuthUser;
use crate::error::AppError;
use crate::flows::{service, FlowInput, RenameInput};
use crate::state::AppState;

/// `POST /api/v1/flows` — save a new flow for the caller.
pub async fn create(
    State(state): State<AppState>,
    user: AuthUser,
    Json(input): Json<FlowInput>,
) -> Result<(StatusCode, Json<Value>), AppError> {
    let flow = service::create(&state.db, &user.id, input).await?;
    Ok((
        StatusCode::CREATED,
        Json(json!({ "status": "success", "data": flow })),
    ))
}

/// `GET /api/v1/flows` — list the caller's flows as summaries (no graph),
/// most recently updated first.
pub async fn list(State(state): State<AppState>, user: AuthUser) -> Result<Json<Value>, AppError> {
    let flows = service::list(&state.db, &user.id).await?;
    Ok(Json(
        json!({ "status": "success", "data": { "flows": flows } }),
    ))
}

/// `GET /api/v1/flows/{id}` — open one of the caller's flows (full graph).
pub async fn get(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, AppError> {
    let flow = service::get(&state.db, &user.id, id).await?;
    Ok(Json(json!({ "status": "success", "data": flow })))
}

/// `PUT /api/v1/flows/{id}` — re-save the caller's open flow (name + graph).
pub async fn update(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
    Json(input): Json<FlowInput>,
) -> Result<Json<Value>, AppError> {
    let flow = service::update(&state.db, &user.id, id, input).await?;
    Ok(Json(json!({ "status": "success", "data": flow })))
}

/// `PATCH /api/v1/flows/{id}` — rename the caller's flow; returns the updated
/// summary (the graph is untouched).
pub async fn rename(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
    Json(input): Json<RenameInput>,
) -> Result<Json<Value>, AppError> {
    let flow = service::rename(&state.db, &user.id, id, input).await?;
    Ok(Json(json!({
        "status": "success",
        "data": {
            "id": flow.id,
            "name": flow.name,
            "created_at": flow.created_at,
            "updated_at": flow.updated_at,
        }
    })))
}

/// `DELETE /api/v1/flows/{id}` — delete the caller's flow; echoes the id.
pub async fn delete(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, AppError> {
    service::delete(&state.db, &user.id, id).await?;
    Ok(Json(json!({ "status": "success", "data": { "id": id } })))
}
